core/controllers/reader_test.py

```diff
@@ -2594,3 +2609,177 @@ def test_submit_learner_answer_details_for_question(self):
+        self.put_json(
+            '/explorehandler/checkpoint_reached/%s' % exp_id,
+            {
+                'most_recently_reached_checkpoint_exp_version': 2,
+                'most_recently_reached_checkpoint_state_name': 'What language'
+            }
+        )
+        exploration_user_data = exp_fetchers.get_exploration_user_data(
+            viewer_id, exp_id)
+        self.assertEqual(
+            exploration_user_data.furthest_reached_checkpoint_exp_version, 2)
+        self.assertEqual(
+            exploration_user_data.furthest_reached_checkpoint_state_name,
+            'What language')
+        self.assertEqual(
+            exploration_user_data.most_recently_reached_checkpoint_exp_version,
+            2)
+        self.assertEqual(
+            exploration_user_data.most_recently_reached_checkpoint_state_name,
+            'What language')
+
+        self.logout()
+
+
+class ExplorationRestartEventHandlerTests(test_utils.GenericTestBase):
+    """Tests for exploration restart event handler."""
+
+    def test_user_checkpoint_progress_is_updated_correctly_on_restart(self):
+        self.signup(self.OWNER_EMAIL, self.OWNER_USERNAME)
+        self.signup(self.VIEWER_EMAIL, self.VIEWER_USERNAME)
+
+        # Load demo exploration.
+        exp_id = '0'
+        exp_services.delete_demo('0')
+        exp_services.load_demo('0')
+
+        self.login(self.VIEWER_EMAIL)
+        viewer_id = self.get_user_id_from_email(self.VIEWER_EMAIL)
+
+        # Viewer opens exploration.
+        exploration_dict = self.get_json(
+            '%s/%s' % (feconf.EXPLORATION_INIT_URL_PREFIX, exp_id))
+        self.assertIsNone(
+            exploration_dict['furthest_reached_checkpoint_exp_version'])
+        self.assertIsNone(
+            exploration_dict['furthest_reached_checkpoint_state_name'])
+        self.assertIsNone(
+            exploration_dict['most_recently_reached_checkpoint_exp_version'])
+        self.assertIsNone(
+            exploration_dict['most_recently_reached_checkpoint_state_name'])
+
+        # First checkpoint reached.
+        self.put_json(
+            '/explorehandler/checkpoint_reached/%s' % exp_id,
+            {
+                'most_recently_reached_checkpoint_exp_version': 1,
+                'most_recently_reached_checkpoint_state_name': 'Welcome!'
+            }
+        )
+        exploration_user_data = exp_fetchers.get_exploration_user_data(
+            viewer_id, exp_id)
+        self.assertEqual(
+            exploration_user_data.furthest_reached_checkpoint_exp_version, 1)
+        self.assertEqual(
+            exploration_user_data.furthest_reached_checkpoint_state_name,
+            'Welcome!')
+        self.assertEqual(
+            exploration_user_data.most_recently_reached_checkpoint_exp_version,
+            1)
+        self.assertEqual(
+            exploration_user_data.most_recently_reached_checkpoint_state_name,
+            'Welcome!')
+
+        # Exploration restarted.
```

Comment:
When the restart button is clicked, following events occur in order:
1. A POST request is send to `explorehandler/restart/<exploration_id>` that sets 'most_recently_reached_checkpoint' to None.
2. The page reloads, due to which a GET request is sent to `explorehandler/init/<exploration_id>`.

The exploration flow from then onwards is same as when an exploration is started for the first time. Since the 'most_recently_reached_checkpoint' is None in both cases, exploration starts from the first state. That being a checkpoint, sends a POST request to `explorehandler/checkpoint_reached/<exploration_id>` that updates the learner checkpoint progress marking the 'first state' as the 'most_recently_reached_checkpoint'.